### ISSUE TITLE
Turning off a host has different behavior on sleeping actors and computing actors

### ISSUE THREAD

**Issue by henricasanova** [simgrid/simgrid#325]
When turning off a host that runs an actor that is in the middle of sleeping the behavior seems to be:
  - The actor does not get an exception in the sleep, doesn't return from main, and its on_exit() function is called

When turning off a host that runs an actor that is in the middle of computing the behavior seems to be:
  - The actor =get an exception in the execute, it can return from main, and its on_exit() function is called

I am attaching a MWE that shows this behavior.  The above may sounds nitkpicky, but as we are writing a whole system on top of SimGrid, it's really inconvenient that the behavior depends on what the actor is doing (especially because in our system sometimes we simulate computation as sleeps! :)). 

[mwe_killing_computing_and_sleeping_actors.zip](https://github.com/simgrid/simgrid/files/2911906/mwe_killing_computing_and_sleeping_actors.zip)

**Comment by mquinson** [comment]
I have a fix for this one, commited locally, but it breaks some MSG tests so it will have to wait tomorrow.

**Comment by henricasanova** [comment]
No problem. While I am at it, if a host is turned off while an actor on that host does a Mailbox::get(), then the behavior is the same as that for a sleep(): the actor doesn't not catch the HostError exception, just dies, and its on_exit() function is called.

**Comment by mquinson** [comment]
This is the expected behavior, I believe :)

**Comment by henricasanova** [comment]
Really? So if I am stuck waiting for a message and my host shuts down, then I don't get told "your host died"? that seems  bit counter-intuitive no?  Or are you saying that I would get a NetworkError because one of the end-points died (in this case me)?

**Comment by mquinson** [comment]
That's a design issue, and I keep bouncing between opinions. If we kill the actor right away, the users cannot do any cleanups without relying on on_exit(), which may be cumbersome. If we don't, people will survive the host failures by accident, and I'm not sure we want to have actors not dying when the host are turned off, do we?

We could add a callback to all those functions specifying what to do if the host is turned off.
Or maybe I'm just paranoid about the actors surviving their host, and we should trust the users to only do what they really mean. 

Maybe some sort of "static cleaners" like this could help:
```
void myfunc() {
  void *payload = malloc(512);
  sg::static_cleaner cleaner([&payload](){
     free(payload);
  });

  mb->put(payload);
  this actor->execute(42);
}
```

The callback of cleaner would be called if (and only if) the actor gets brutally killed before this variable gets out of scope.

Feedback welcome...

**Comment by henricasanova** [comment]
What you're talking about above is exactly the issues I've been dealing with today.  I was initially thinking of having a massive try-catch around the whole main() to catch HostError, as a way to handle the "my host died" event and do all necessary cleanup (that then cannot place any simgrid calls, which is fine). In other words, I was behaving like a good user who doesn't forget to handle the host dying.  But that didn't work because of the above design decision, which as you say can change.  Then I thought: whatever, I'll have a "has_exited_cleanly" flag that's false and only set to true when main() returns. That way, in the on_exit function I can tell whether this is a normal exit or a "my host has failed" exit. Then I can take appropriate actions.  But as you say, it's a bit cumbersome (in my case with a boolean passed to the on_exit function, in your case with a special cleaner that is called upon brutal death, but that's the same thing really). 

So, to summarize: "I agree" :)  

Now, if you throw the HostError, won't the actor die with some "uncaught HostError"? Thus preventing the user from mistakenly continuing doing things without noticing the HostError?  Perhaps I am confused now...

**Comment by mquinson** [comment]
I'm concerned about users who catch HostError but don't stop their actor in that case. I fear that other calls may segfault if this situation occurs, and also I'm not sure that this is the "right thing to do".

Buy maybe I should just trust the users and give them tools to do the necessarily cleanups. Btw, what do you think of my proposal for a static cleaner in functions ?

**Comment by henricasanova** [comment]
Oh I see, you're concerned about the users catching the HostError exception but, e.g., doing nothing in the catch, and I confirm that segfaults will occur (and good luck using valgrind/debugger to figure it out... it basically says something about some invalid instruction I think).  

I like your cleaner idea, but in my case, I have to do several things whenever there is a brutal crash that are not about freeing memory because references get out of scope (I do some state updates to objects,  etc).  So what I would need is more of a general "you crashed" callback.  How about an "on_exit" function (that gets called when the actor returns from main, and gets the return code), and a an "on_died" function (that gets called when the host was shutdown or the actor was killed by somebody...i.e., whenever the actor no longer run and didn't return by itself from main).

**Comment by mquinson** [comment]
We could also add a parameter to the callbacks of on_exit(), specifying whether it was a forceful exit or not. That's actually what we had in MSG, but I removed that extra parameter because I was never using it in my (tiny) examples.

Another solution would be to make the StopRequest exception visible from the users (if it's not already). That way, if you want to cleanup stuff, you'd have to catch StopRequest, do your cleanups, and rethrow it. I'd be specified in the doc that catching StopRequest without rethrowing it will probably lead the simulation to segfault. This exception could be renamed ForcefulKill or SIGKILL or a nice name implying "don't mess with that exception, bro".

**Comment by henricasanova** [comment]
> We could also add a parameter to the callbacks of on_exit(), specifying whether it was a forceful exit or not. That's actually what we had in MSG, but I removed that extra parameter because I was never using it in my (tiny) examples.

Yes. That's essentially what I hacked up on my side, where in on_exit() I check "forceful or not" and if not forceful, I get what main returned.  That, in addition to your cleaners to avoid memory leaks would probably work. And yes, I never started caring about this until implemented pretty large multi-actor stuff and aiming for fault-tolerance/recovery.


> Another solution would be to make the StopRequest exception visible from the users (if it's not already). That way, if you want to cleanup stuff, you'd have to catch StopRequest, do your cleanups, and rethrow it. I'd be specified in the doc that catching StopRequest without rethrowing it will probably lead the simulation to segfault. This exception could be renamed ForcefulKill or SIGKILL or a nice name implying "don't mess with that exception, bro".

Perhaps... Like you I go back and forth between the two: (i) you're not careful enough to catch/rethrow-or-return and so you just will find yourself in on_exit and that's it; and (ii) we trust you to be careful. 

In my particular case, the first solution works for everything I need. The second solution would allow more fine-grain control (i.e., depending on where in your code you get the exception you'd do different things...).  

For now, I hacked up WRENCH to do Solution 1 above outside of SimGrid with the master branch as it is, and it works. So perhaps we can table this issue until we see each other f2f in a few months... and by then we might have converged?

**Comment by mquinson** [comment]
Actually, since changing the semantic of released code is never good, I'd love if we could come up with a solution before the next release, by the end of March...

**Comment by henricasanova** [comment]
Forgot about the release. Then I am thinking Solution #1. After all, if you get killed, you shouldn't run code in your main(). So I think that's the most intuitive.  Do now, whenever you're killer and/or your host goes down, you never get the ExecutionError exception (i.e., not while you compute, sleep, put, or get). Instead, you just find yourself in on_exit, which tells you whether you were brutally terminated or not, and if not what main returned. 

sounds good?

**Comment by henricasanova** [comment]
I just did a small example with  sleeping, computing, receiving, and sending actors, and it all behaves as we discussed above. So I believe it's ready for closing. I'll let Martin do it as he may still be doing stuff here.

**Comment by henricasanova** [comment]
Ok, I came up with another MWE that breaks things a bit.  In the attached code, you'll see that a Receiver that is on a host that is turned off will actually run the code in the catch of the NetworkFailureException (amusingly, it seems to be doing that AFTER its on_exit() function is called!).  It's as if the on_exit() function turns to where the exception was caught..... 

[mwe_killing_endpoint_leads_to_network_exception_at_endpoint.zip](https://github.com/simgrid/simgrid/files/2928495/mwe_killing_endpoint_leads_to_network_exception_at_endpoint.zip)

**Comment by mquinson** [comment]
I'm glad you find it amusing :)

On my side, I'm done implementing the solution 1, ie, cleaning up what the first parameter of on_exit callbacks means (spoiler: it should be named 'failed'), and also implementing the solution 2, ie, making the StopRequest exception public, under the name ForcefulKillException. I finish my tests and push.

Later (tomorrow at least), I will improve/write the doc of S4U wrt exception, and fix your new bug.

**Comment by mquinson** [comment]
Hello,

next time we have an issue in this area, it'd be really great if you could add a new test to teshsuite/s4u/activity-lifecycle/activity-lifecycle.cpp as a MWE. It'd save us a conversion :)

Thanks in advance,
Mt

**Comment by henricasanova** [comment]
Sure. Do you want me to do it now for this MWE?

**Comment by mquinson** [comment]
That would be really cool because I'm really struggling finding time to code in simgrid. 

Actually, I already integrated a MWE for #323 in this code (function test_sleep_restart_end), and it accurately reproduce the failure reported, but I failed to find the time to investiguate further... So even with a MWE, I really hope that you have a workaround on your side for this problem :( 

Sorry,

**Comment by henricasanova** [comment]
No workaround possible here, but I am working on a development branch that uses the SimGrid master branch, planning to merge it into the WRENCH master after SimGrid 3.22 has been release. So there is no huge time pressure, but of course the earlier the better since right now I can't test whether WRENCH handles host failures correctly.  I'll work on the tesh conversion soon.

**Comment by henricasanova** [comment]
I wasn't sure what the procedure was to add a test that fails to the teshsuite, so I've commented out the call to the test (test is called "test_host_off_while_receive").

**Comment by henricasanova** [comment]
I believe this has been fixes. @mquinson  do you confirm? It all works on my end now. I'll let you close this?

**Comment by mquinson** [comment]
This test was integrated, but it's still commented. I just tested, and it still fails. So we cannot close this issue.

**Comment by henricasanova** [comment]
The test fails in a different way now after the fixed you made!  (Which makes sense as WRENCH now works correctly). This is actually a but in the test (which now passes the fail case, and then encounter that stupid bug...). I just fixed it and the behavior is correct.  I'd like to push to framagit, but I am not allowed. Can you add me to the repo as a contributor? Once I've pushed I'll close this issue.
